src/mozilla_sec_eia/ex_21/inference.py

```diff
@@ -0,0 +1,342 @@
+    """Perform basic cleaning on a dataframe extracted from an Ex. 21."""
+    if extracted_df.empty:
+        return extracted_df
+    if "row" in extracted_df.columns:
+        extracted_df = extracted_df.drop(columns=["row"])
+    extracted_df["subsidiary"] = extracted_df["subsidiary"].str.strip().str.lower()
+    if "loc" in extracted_df.columns:
+        extracted_df["loc"] = extracted_df["loc"].str.strip().str.lower()
+    if "own_per" in extracted_df.columns:
+        # could do some special character removal on all columns?
+        extracted_df["own_per"] = extracted_df["own_per"].str.replace(
+            r"[^\w.]", "", regex=True
+        )
+    return extracted_df
+
+
+def _sort_by_label_priority(target_array):
+    _, label2id = get_id_label_conversions(LABELS)
+    id_priority = [label2id[label] for label in LABEL_PRIORITY]
+    # Create a priority map from the label priority
+    priority_map = {val: idx for idx, val in enumerate(id_priority)}
+    # Sort the target array based on the priority map
+    sorted_array = sorted(target_array, key=lambda x: priority_map.get(x, float("inf")))
+    return sorted_array
+
+
+def get_flattened_mode_predictions(token_boxes_tensor, predictions_tensor):
+    """Get the mode prediction for each box in an Ex. 21.
+
+    When handling multi page documents LayoutLM uses a sliding 'frame'
+    with some overlap between frames. The overlap creates multiple
+    predictions for the same bounding boxes. Thus it's necessary to find
+    the mode of all the predictions for a bounding box and use that as the
+    single prediction for each box. If there are multiple mode
+    predictions for a bounding box, then ties are broken by setting
+    a priority for the labels (LABEL_PRIORITY) and choosing the highest priority
+    label.
+    """
+    # Flatten the tensors
+    flat_token_boxes = token_boxes_tensor.view(-1, 4)
+    flat_predictions = predictions_tensor.view(-1)
+
+    boxes = flat_token_boxes.numpy()
+    predictions = flat_predictions.numpy()
+
+    # Find unique boxes and indices
+    unique_boxes, inverse_indices = np.unique(boxes, axis=0, return_inverse=True)
+
+    # Compute the mode for each unique bounding box
+    # for each unique box in boxes, create a list with all predictions for that box
+    # get the indices in predictions where the corresponding index in boxes is
+    unique_box_predictions = [
+        predictions[np.where(inverse_indices == i)[0]] for i in range(len(unique_boxes))
+    ]
+    pred_counts = [np.bincount(arr) for arr in unique_box_predictions]
+    # Compute the mode of predictions for each group
+    # break ties by taking into account LABEL_PRIORITY
+    modes = np.array(
+        [
+            _sort_by_label_priority(np.where(arr == np.max(arr))[0])[0]
+            for arr in pred_counts
+        ]
+    )
+    flattened_modes = modes[inverse_indices]
+
+    return flattened_modes
+
+
+def _get_data(dataset):
+    yield from dataset
+
+
+def perform_inference(
```

Comment:
This function does the work in this module. It starts by creating a HuggingFace dataset from PDFs in `pdfs_dir` that the `model` can then perform inference on (`create_inference_dataset`). Then it creates an instance of the inference pipeline and runs the dataset through the pipeline. The pipeline outputs logits, predictions, and an output dataframe with extracted Ex. 21 table.